src/null-safety/index.md

```diff
@@ -1,41 +1,288 @@
+| `Map<String, int?>?` | {{yes}}                       | {{yes}}                 |
+{:.table .table-striped}
+
+_\* Even when all the int values in the map are non-null,
+when you use an invalid key to do a map lookup, the returned value is null._
+
+Because map lookups can return null,
+you can't assign them to non-nullable variables:
+
+```dart
+// Assigning a lookup result to a non-nullable
+// variable causes an error.
+int value = <String, int>{'one': 1}['one']; // ERROR
+```
+
+{% comment %}
+**[QUESTION: Will the analyzer ever be able to detect that is non-null?]**
+{% endcomment %}
+
+One workaround is to change the type of the variable to be nullable:
+
+```dart
+int? value = <String, int>{'one': 1}['one']; // OK
+```
+
+Another way to fix the problem —
+if you're sure the lookup succeeds —
+is to add a `!`:
+
+```dart
+int value = <String, int>{'one': 1}['one']!; // OK
+```
+
+A safer approach is to use the lookup value only if it's not null.
+You can test its value using
+an [`if` statement][] or the [`??` operator][].
+Here's an example of using the value `0` if the lookup returns a null value:
+
+```dart
+var aList = <String, int>{'one': 1};
+...
+int value = aList['one'] ?? 0;
+```
+
+
+## Fixing analysis errors
+
+With null safety, the Dart analyzer generates errors when
+it finds a nullable value where a non-null value is required.
+That's not as bad as it sounds:
+the analyzer can often recognize when
+a variable or expression inside a function has
+a nullable type but can't have a null value.
+However, the analyzer can't model the flow of your whole application,
+so it can't predict the value of a global variable or class field.
+
+If you know a value can't be null
+but it has a nullable type,
+consider using `!`, `late`, or `as`.
+Examples of using `!` and `late` are in the [first section of this page][].
+Here's an example of using `as` to convert a `num?` to a `int`:
+
+```dart
+return maybeNum() as int;
+```
+
+[first section of this page]: #creating-and-using-variables---late
+
+If a value might indeed be null,
+then you can explicitly check for that.
+A common approach is to use an `if` statement or `??` expression
+to check for and handle a null value.
+Dart then considers the code for the non-null value to be null safe.
+Here's an example of using `??`:
+
+```dart
+int value = aNullableInt ?? 0; // 0 if it's null; otherwise, the integer
+```
+
+Here's an example of using an `if` statement:
+
+```dart
+int definitelyInt(int? aNullableInt) {
+  if (aNullableInt == null) {
+    return 0;
+  }
+  return aNullableInt; // Can't be null!
+}
+```
+
+## Other common code changes
+
+Once you opt into null safety,
+you can't use the [member access operator (`.`)][other operators]
+if the operand might be null.
+Instead, you can use the null-aware version of that operator (`?.`):
+
+```dart
+double? d;  
+print(d?.floor()); // Uses `?.` instead of `.` to invoke `floor()`.
+```
+
+
+## Where to learn more
```

Comment:
I think we need some details on how to actually run with null safety, either in the text above, or in a link here.

Reply:
We could consider just linking to this? https://github.com/dart-lang/samples/blob/master/null_safety/calculate_lix/README.md